cmd/thanos/receive.go

```diff
@@ -508,6 +516,22 @@ func setupHashring(g *run.Group,
 					return nil
 				}
 				webHandler.Hashring(h)
+
+				// Because hashring configs from hashrings file can change,
```

Comment:
Can we send validated `[]receiver.HashringConfigs` on the channel instead of the `Hashring`? This way we won't have to do another validation when we receive the hashring.

Reply:
Now looking at this more closely, I think keeping the channel of `Hashring` is better.

The channel is not just used for `MultiTSDB`. It's used for [Handler](https://github.com/thanos-io/thanos/pull/5777/files#diff-3d45f9beb5b670e94fe2715e21477a3fb8f5db7bf97f01cddf6e108bb98761a6L510) too. And `Handler` needs `Hashring`, not `[]HashringConfig`.

If we change to channel of `[]HashringConfig`, we still need to set hashring from it for `Handler`. In this case we have three cases to check: config from hashring file, config from not empty yaml content, config from empty yaml content.

That's even bigger validation than now.

Does it make sense?

Reply:
Yes, that's the case of not empty yaml content, where hashring can't change.

As you can see [here](https://github.com/thanos-io/thanos/blob/95430de86da3d5d1017bb0eff722fc68ab092588/cmd/thanos/receive.go#L451-L492), we have three cases:
- Config from hashring file (changeable): `HashringFromConfigWatcher`
- Config from not empty yaml content (unchangeable): `NewMultiHashring` (this can not be used for changeable config)
- Config from empty yaml content: `SingleNodeHashring`

If we set the channel of `[]HashringConfig`, we will have to check for these three cases and set hashring for `Handler`

Reply:
Actually, let me rethink^^

Reply:
Yeah, now rethought, I think we only have two cases:
- `NewMultiHashring` for normal cases
- [SingleNodeHashring when config content is empty](https://github.com/thanos-io/thanos/blob/afdb30e048d0dfb0de3764dcc73164c6589cb894/cmd/thanos/receive.go#L483)

Because this channel receives both the initial hashring and the changed hashring, we have to deal with both these cases.

So basically this's still a bigger validation than what we have now. So I still keep my point of keeping the channel of `Hashring`.

Reply:
Actually, let me rethink...

Reply:
Yeah, let me try setting the channel of `[]HashringConfig`. We may still need a validation for empty content. But that validation would probably be simpler than the current validation.

Reply:
I did this in the last commit. There're some failed e2e tests. Could you check if the failed tests are related to this new change?

I think they are. I saw a few "no matching hashring to handle tenant" in the logs.

Reply:
Wait, I may be seeing what's wrong here

Reply:
Yeah e2e are fine now